File: nodejs/src/client/approvals.ts
/**
 * Approvals API — list pending approval requests and resolve them (approve/reject).
 * The same decision the web inbox and Slack/email buttons make, resolving the shared
 * approval record. Backend: agent-orchestration approval-requests REST API.
 *
 *   GET  /agent-orchestration/organizations/{org}/approval-requests?status=pending
 *   POST /agent-orchestration/organizations/{org}/approval-requests/{id}/approve|reject
 */

import type { LoxtepHttpClient } from '../http/client.js';
import type {
  ApprovalsApiDeps,
  ApprovalsListFilters,
  ApprovalsListResponse,
  ApprovalsResolveOptions,
  ApprovalDecisionResult,
} from './approvals-types.js';

function unwrap<T>(res: unknown): T {
  return ((res as { data?: T }).data ?? res) as T;
}

function normalizeResolveOptions(
  organization_idOrOptions?: string | ApprovalsResolveOptions
): ApprovalsResolveOptions {
  if (typeof organization_idOrOptions === 'string') {
    return { organization_id: organization_idOrOptions };
  }
  return organization_idOrOptions ?? {};
}

function buildResolveBody(options: ApprovalsResolveOptions): Record<string, unknown> {
  const body: Record<string, unknown> = {};
  if (options.response !== undefined) body.response = options.response;
  if (options.form_response !== undefined) body.form_response = options.form_response;
  if (options.decision_note !== undefined) body.decision_note = options.decision_note;
  return body;
}

export function createApprovalsApi(
  http: LoxtepHttpClient,
  deps: ApprovalsApiDeps = {}
): {
  list: (filters?: ApprovalsListFilters) => Promise<ApprovalsListResponse>;
  list_pending: (organization_id?: string) => Promise<ApprovalsListResponse>;
  resolve: (
    approval_request_id: string,
    action: 'approve' | 'reject',
    organization_idOrOptions?: string | ApprovalsResolveOptions
  ) => Promise<ApprovalDecisionResult>;
  approve: (
    approval_request_id: string,
    options?: ApprovalsResolveOptions
  ) => Promise<ApprovalDecisionResult>;
  reject: (
    approval_request_id: string,
    options?: ApprovalsResolveOptions
  ) => Promise<ApprovalDecisionResult>;
} {
  const resolveOrg = (override?: string): string => {
    const org = override ?? deps.organization_id;
    if (!org) {
      throw new Error(
        'organization_id is required for approvals calls (set it on the client or pass it explicitly)'
      );
    }
    return org;
  };

  const base = (org: string) =>
    `/agent-orchestration/organizations/${encodeURIComponent(org)}/approval-requests`;

  return {
    async list(filters: ApprovalsListFilters = {}): Promise<ApprovalsListResponse> {
      const org = resolveOrg(filters.organization_id);
      const search = new URLSearchParams();
      if (filters.status) search.set('status', filters.status);
      if (filters.page != null) search.set('page', String(filters.page));
      if (filters.page_size != null) search.set('page_size', String(filters.page_size));
      const qs = search.toString() ? `?${search.toString()}` : '';
      const res = await http.get<ApprovalsListResponse>(`${base(org)}${qs}`);
      const payload = unwrap<ApprovalsListResponse>(res);
      const out: ApprovalsListResponse = {
        items: payload.items ?? [],
        pagination: payload.pagination,
      };
      if (payload.total != null) out.total = payload.total;
      return out;
    },

    list_pending(organization_id?: string): Promise<ApprovalsListResponse> {
      return this.list({ status: 'pending', organization_id });
    },

    async resolve(
      approval_request_id: string,
      action: 'approve' | 'reject',
      organization_idOrOptions?: string | ApprovalsResolveOptions
    ): Promise<ApprovalDecisionResult> {
      const options = normalizeResolveOptions(organization_idOrOptions);
      const org = resolveOrg(options.organization_id);
      const res = await http.post<ApprovalDecisionResult>(
        `${base(org)}/${encodeURIComponent(approval_request_id)}/${action}`,
        buildResolveBody(options)
      );
      return unwrap<ApprovalDecisionResult>(res);
    },

    approve(
      approval_request_id: string,
      options?: ApprovalsResolveOptions
    ): Promise<ApprovalDecisionResult> {
      return this.resolve(approval_request_id, 'approve', options);
    },

    reject(
      approval_request_id: string,
      options?: ApprovalsResolveOptions
    ): Promise<ApprovalDecisionResult> {
      return this.resolve(approval_request_id, 'reject', options);
    },
  };
}

File: python/src/loxtep/approvals.py
"""
Approvals API — list pending approval requests and resolve them (approve/reject).
Backend: agent-orchestration approval-requests REST API.

  GET  /agent-orchestration/organizations/{org}/approval-requests?status=pending
  POST /agent-orchestration/organizations/{org}/approval-requests/{id}/approve|reject
"""

from __future__ import annotations

from typing import Any, Optional
from urllib.parse import urlencode

from .http_client import AsyncLoxtepHttpClient, LoxtepHttpClient


def _unwrap(res: Any) -> Any:
    return res.get("data", res) if isinstance(res, dict) else res


def _query_string(params: dict[str, Any]) -> str:
    clean = {k: str(v) for k, v in params.items() if v is not None}
    return "?" + urlencode(clean) if clean else ""


class ApprovalsApi:
    """Sync approvals surface (HITL inbox parity)."""

    unavailable: bool = False

    def __init__(
        self,
        http: LoxtepHttpClient,
        *,
        organization_id: Optional[str] = None,
    ) -> None:
        self._http = http
        self._organization_id = organization_id

    def _resolve_org(self, override: Optional[str] = None) -> str:
        org = override or self._organization_id
        if not org:
            raise ValueError(
                "organization_id is required for approvals calls "
                "(set it on the client or pass it explicitly)"
            )
        return org

    def _base(self, org: str) -> str:
        return f"/agent-orchestration/organizations/{org}/approval-requests"

    def list(
        self,
        *,
        status: Optional[str] = None,
        organization_id: Optional[str] = None,
        page: Optional[int] = None,
        page_size: Optional[int] = None,
    ) -> dict[str, Any]:
        org = self._resolve_org(organization_id)
        qs = _query_string({"status": status, "page": page, "page_size": page_size})
        payload = _unwrap(self._http.get(f"{self._base(org)}{qs}"))
        if not isinstance(payload, dict):
            return {"items": [], "pagination": None}
        return {
            "items": payload.get("items") or [],
            "pagination": payload.get("pagination"),
        }

    def list_pending(self, organization_id: Optional[str] = None) -> dict[str, Any]:
        return self.list(status="pending", organization_id=organization_id)

    def resolve(
        self,
        approval_request_id: str,
        action: str,
        organization_id: Optional[str] = None,
    ) -> dict[str, Any]:
        if action not in ("approve", "reject"):
            raise ValueError("action must be 'approve' or 'reject'")
        org = self._resolve_org(organization_id)
        return _unwrap(
            self._http.post(f"{self._base(org)}/{approval_request_id}/{action}", {})
        )

    def approve(
        self,
        approval_request_id: str,
        organization_id: Optional[str] = None,
    ) -> dict[str, Any]:
        return self.resolve(approval_request_id, "approve", organization_id)

    def reject(
        self,
        approval_request_id: str,
        organization_id: Optional[str] = None,
    ) -> dict[str, Any]:
        return self.resolve(approval_request_id, "reject", organization_id)


class AsyncApprovalsApi:
    """Async approvals surface (HITL inbox parity)."""

    unavailable: bool = False

    def __init__(
        self,
        http: AsyncLoxtepHttpClient,
        *,
        organization_id: Optional[str] = None,
    ) -> None:
        self._http = http
        self._organization_id = organization_id

    def _resolve_org(self, override: Optional[str] = None) -> str:
        org = override or self._organization_id
        if not org:
            raise ValueError(
                "organization_id is required for approvals calls "
                "(set it on the client or pass it explicitly)"
            )
        return org

    def _base(self, org: str) -> str:
        return f"/agent-orchestration/organizations/{org}/approval-requests"

    async def list(
        self,
        *,
        status: Optional[str] = None,
        organization_id: Optional[str] = None,
        page: Optional[int] = None,
        page_size: Optional[int] = None,
    ) -> dict[str, Any]:
        org = self._resolve_org(organization_id)
        qs = _query_string({"status": status, "page": page, "page_size": page_size})
        payload = _unwrap(await self._http.get(f"{self._base(org)}{qs}"))
        if not isinstance(payload, dict):
            return {"items": [], "pagination": None}
        return {
            "items": payload.get("items") or [],
            "pagination": payload.get("pagination"),
        }

    async def list_pending(self, organization_id: Optional[str] = None) -> dict[str, Any]:
        return await self.list(status="pending", organization_id=organization_id)

    async def resolve(
        self,
        approval_request_id: str,
        action: str,
        organization_id: Optional[str] = None,
    ) -> dict[str, Any]:
        if action not in ("approve", "reject"):
            raise ValueError("action must be 'approve' or 'reject'")
        org = self._resolve_org(organization_id)
        return _unwrap(
            await self._http.post(f"{self._base(org)}/{approval_request_id}/{action}", {})
        )

    async def approve(
        self,
        approval_request_id: str,
        organization_id: Optional[str] = None,
    ) -> dict[str, Any]:
        return await self.resolve(approval_request_id, "approve", organization_id)

    async def reject(
        self,
        approval_request_id: str,
        organization_id: Optional[str] = None,
    ) -> dict[str, Any]:
        return await self.resolve(approval_request_id, "reject", organization_id)
